core/src/main/java/org/infinispan/interceptors/locking/AbstractLockingInterceptor.java

```diff
@@ -80,14 +82,13 @@ public Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryComm
    protected abstract Object visitDataWriteCommand(InvocationContext ctx, DataWriteCommand command) throws Throwable;
 
    // We need this method in here because of putForExternalRead
-   protected Object visitNonTxDataWriteCommand(InvocationContext ctx, DataWriteCommand command) throws Throwable {
+   protected final Object visitNonTxDataWriteCommand(InvocationContext ctx, DataWriteCommand command) throws Throwable {
       try {
-         if (!shouldLock(command.getKey(), command))
+         if (hasSkipLocking(command) || !shouldLockKey(command.getKey())) {
             return invokeNextInterceptor(ctx, command);
-         lockKey(ctx, command);
+         }
+         lockAndRecord(ctx, command.getKey(), getLockTimeoutMillis(command));
          return invokeNextInterceptor(ctx, command);
-      } catch (Throwable te) {
-         throw cleanLocksAndRethrow(ctx, te);
```

Comment:
Yeah but they don't override the method. It seems they should be doing a try finally block like we have here instead to me.